File: apps/studio/backend/app/features/execution_resolution/schemas.py
"""Public contracts for exact executable identity resolution."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.features.telemetry_contract.types import NonEmptyPortableText, PortableText

ExecutableType = Literal["workflow", "subflow", "agent"]


class ExecutionResolutionQuery(BaseModel):
    model_config = ConfigDict(extra="forbid")

    service_namespace: PortableText = Field(
        description="Exact service.namespace; empty is explicit"
    )
    service_name: NonEmptyPortableText = Field(description="Exact service.name")
    executable_type: ExecutableType
    runtime_id: NonEmptyPortableText


class ExecutionResolution(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    service_namespace: PortableText
    service_name: NonEmptyPortableText
    executable_type: ExecutableType
    runtime_id: NonEmptyPortableText
    trace_id: str = Field(pattern="^[0-9a-f]{32}$")
    span_id: str = Field(pattern="^[0-9a-f]{16}$")
    detail_path: str = Field(pattern="^/")
    trace_path: str = Field(pattern="^/")


class ExecutionResolutionConflictResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    code: Literal["ambiguous_execution_identity"]
    message: NonEmptyPortableText
    match_count: int = Field(ge=2)

File: apps/studio/frontend/src/features/execution-resolution/schemas.ts
import { z } from 'zod'

export const ExecutableTypeSchema = z.enum(['workflow', 'subflow', 'agent'])
export type ExecutableType = z.infer<typeof ExecutableTypeSchema>

export const ExecutionResolutionRequestSchema = z.object({
  service_namespace: z.string(),
  service_name: z.string().min(1),
  executable_type: ExecutableTypeSchema,
  runtime_id: z.string().min(1),
  destination: z.enum(['detail', 'trace']),
}).strict()
export type ExecutionResolutionRequest = z.infer<typeof ExecutionResolutionRequestSchema>

export const ExecutionResolutionSchema = z.object({
  service_namespace: z.string(),
  service_name: z.string().min(1),
  executable_type: ExecutableTypeSchema,
  runtime_id: z.string().min(1),
  trace_id: z.string().regex(/^[0-9a-f]{32}$/),
  span_id: z.string().regex(/^[0-9a-f]{16}$/),
  detail_path: z.string().startsWith('/'),
  trace_path: z.string().startsWith('/'),
}).strict()
export type ExecutionResolution = z.infer<typeof ExecutionResolutionSchema>

export const ExecutionResolutionConflictSchema = z.object({
  code: z.literal('ambiguous_execution_identity'),
  message: z.string().min(1),
  match_count: z.number().int().min(2),
}).strict()
